### ISSUE TITLE
memory usage

### ISSUE THREAD

**Issue by tromp** [AidosKuneen/cuckoo#3]
I noticed your miner is said to use 128 MB, whereas my mean25 miner uses under 70 MB.
Is this because the C code can use unaligned accesses to store edges in only 40 bits, and the Go code cannot?

**Comment by ogami-daigoro** [comment]
As you said for now in our code edges are not stored  in 40 bits since our number of nodes
is small so memory usage is not a big issue. 
And also because  we want to make codes simpler.
(I understand the fully optimized C code is a little bit complicated for  the smaller storage.)

**Comment by tromp** [comment]
Thinking about it, the smaller size EDGEBITS=24 doesn't need 40-bit encodings. 32-bits suffices
for the initial rounds. After that, you have enough space to even use 64 bit edges.
So with some care, the Go solver can be rewritten to use under 70 MB as well...

**Comment by ogami-daigoro** [comment]
I'm working on another code and almost forgot about  cuckoo code.
Give me some time for the answer.

**Comment by ogami-daigoro** [comment]
First of all my code is not a complete copy of your mean miner.

I understand that your mean miner  has a 2 dimension matrix [x][y] where length of x and y = 2^5 
and each elements in matrix have (in average) 2^24/((2^5)*(2^5))=2^14 number of 4 bytes data, so (2^5)*(2^5)*(2^14)*4 ≒70MB. And I understand the trick in your code is how we manage (14+24)bits data
for a pair of nodes in 4bytes(32bits) and it uses very sophisticated and awesome but complicated algorithm.

My decision for my code is that   we want speed and my code simpler to maintain longer period for our cryptocurrency. Also we don't mind larger memory usage at certain extent,
since even phones have a few GB of memory. So in our code each elements in matrix  have 8 bytes
instead of 4 bytes.

In short we utilized your excellent idea of reference locality for speed but didn't use memory usage 
optimization.

**Comment by tromp** [comment]
> I understand that your mean miner has a 2 dimension matrix [x][y] where
> length of x and y = 2^5

I use 2^7 in each dimension for my mean30.

> and each matrix has (in average) 2^24/((2^5)(2^5))=2^14 number of 4 bytes
> data, so (2^5)(2^5)*(2^14)*4 ≒70MB.

Yes, this is just NEDGES * sizeof(u32), in your case 2^24*4.

> And I understand the trick in your code
> is how we manage (15+24)bits data
> for u and v nodes in 4bytes(32bits) and it uses very sophisticated and
> awesome but complicated algorithm.
> My decision for my code is that we want speed and my code simpler to
> maintain longer period for our cryptocurrency. Also we don't mind larger
> memory usage at certain extent,
> since even phones has a few GB memory.

Indeed; the GPU miner is much simpler; it turns out that avoiding
the tricky bit manipulations while simply using 64 bits per edge is faster.
Do you plan for GPU mining on your coin as well?

**Comment by ogami-daigoro** [comment]
> Do you plan for GPU mining on your coin as well?

For now no. Our code is for general purpose and not for miners.
